rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+
+This is fast enough for web UI enrollment interactions that might wait for the resource network status to be reported.
+
+The default healthy threshold is set to a higher value than the unhealthy threshold to help avoid false positive `healthy` status when network connectivity is unreliable.
+It makes sense to bias the health status to `unhealthy` when the network is unreliable anyway.
+
+[^2]: https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html
+[^3]: https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-custom-probe-overview
+[^4]: Azure timeouts always match the interval, they cannot be configured independently.
+[^5]: https://docs.docker.com/reference/dockerfile/#healthcheck
+[^6]: https://cloud.google.com/load-balancing/docs/health-check-concepts
+[^7]: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes
+
+#### Dynamic config example
+
+    kind: db
+    version: v3
+    metadata:
+      name: "example"
+      labels:
+        env: "dev"
+    spec:
+      protocol: "postgres"
+      uri: "localhost:5432"
+      health_check:
+        enabled: true
+        interval: 30s
+        timeout: 5s
+        healthy_threshold: 3
+        unhealthy_threshold: 2
+  
+#### Static config example
+
+    db_service:
+      enabled: true
+      databases:
+      - name: "example"
+        protocol: "postgres"
+        uri: "localhost:5432"
+        static_labels:
+          "env": "dev"
+        health_check:
+          enabled: true
+          interval: 30s
+          timeout: 5s
+          healthy_threshold: 3
+          unhealthy_threshold: 2
+
+#### Terraform config example
+
+    resource "teleport_database" "example" {
+      version = "v3"
+      metadata = {
+        name = "example"
+        labels = {
+          "env" = "dev"
+        }
+      }
+      spec = {
+        protocol = "postgres"
+        uri      = "localhost:5432"
+        health_check = {
+          enabled             = true
+          healthy_threshold   = 1
+          unhealthy_threshold = 1
+          interval            = "30s"
+          timeout             = "5s"
+        }
+      }
+    }
+
+#### Discovery config example
+
+Discovery service creates `db` objects and db agents can configure custom health check settings in their dynamic resources matchers.
+
+Dynamic resource matcher config overrides `db.spec.health_check` settings.
+
+    discovery_service:
+      enabled: true
+      discovery_group: "teleport-dev-2"
+      aws:
+        - types:
+            - "rds"
+          regions:
+            - "ca-central-1"
+          tags:
+            "env": "prod"
+
+    db_service:
+      enabled: true
+      resources:
+        - labels:
+            "teleport.dev/origin": "dynamic"
+            "env": "dev"
+          health_check:
+            enabled: false # disable health checks for these databases
```

Comment:
This part will also be not needed on the discovery/db services if health probes are configured as a separate resource which will keep configurations cleaner.